plugins/modules/packaging/os/redhat_subscription.py

```diff
@@ -593,15 +593,22 @@ def update_subscriptions_by_pool_ids(self, pool_ids):
         consumed_pools = RhsmPools(self.module, consumed=True)
 
         existing_pools = {}
+        serials_to_remove = []
         for p in consumed_pools:
-            existing_pools[p.get_pool_id()] = p.QuantityUsed
+            pool_id = p.get_pool_id()
+            quantity_used = p.get_quantity_used()
+            existing_pools[pool_id] = quantity_used
+
+            quantity = pool_ids.get(pool_id, 0)
+            if quantity is not None and quantity != quantity_used:
+                serials_to_remove.append(p.Serial)
 
-        serials_to_remove = [p.Serial for p in consumed_pools if pool_ids.get(p.get_pool_id(), 0) != p.QuantityUsed]
         serials = self.unsubscribe(serials=serials_to_remove)
 
         missing_pools = {}
         for pool_id, quantity in sorted(pool_ids.items()):
-            if existing_pools.get(pool_id, 0) != quantity:
+            quantity_used = existing_pools.get(pool_id, 0)
+            if quantity is None and quantity_used == 0 or quantity is not None and quantity_used != quantity:
```

Comment:
What happens if `quantity == 0`, when `quantity_used > 0`? Then the above loop added the pool ID to `serials_to_remove`, but at the same time we're adding `missing_pools[pool_id] = 0`. That's probably not what should happen, right?

Then this should probably be:
```suggestion
            if quantity is None and quantity_used == 0 or quantity not in (None, 0, quantity_used):
```

Reply:
Agreed and modified.